pkg/agent/core/ngt/service/vqueue/queue.go

```diff
@@ -0,0 +1,326 @@
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if !f(idx.uuid, idx.vector) {
+				return
+			}
+		}
+	}
+}
+
+func (v *vqueue) RangePopDelete(ctx context.Context, f func(uuid string) bool) {
+	if v.finalizing.Load().(bool) {
+		for !v.finalizing.Load().(bool) {
+			time.Sleep(time.Millisecond * 100)
+		}
+	}
+	for _, key := range v.flushAndLoadDelete() {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if !f(key.uuid) {
+				return
+			}
+		}
+	}
+}
+
+func (v *vqueue) GetVector(uuid string) ([]float32, bool) {
+	v.imu.Lock()
+	vec, ok := v.uiil[uuid]
+	v.imu.Unlock()
+	return vec, ok
+}
+
+func (v *vqueue) addInsert(i index) {
+	v.imu.Lock()
+	v.uii = append(v.uii, i)
+	v.uiil[i.uuid] = i.vector
+	v.imu.Unlock()
+}
+
+func (v *vqueue) addDelete(d key) {
+	v.dmu.Lock()
+	v.udk = append(v.udk, d)
+	v.dmu.Unlock()
+}
+
+func (v *vqueue) popInsert() (i index) {
+	v.imu.Lock()
+	i = v.uii[0]
+	v.uii = v.uii[1:]
+	delete(v.uiil, i.uuid)
+	v.imu.Unlock()
+	return i
+}
+
+func (v *vqueue) popDelete() (d key) {
+	v.dmu.Lock()
+	d = v.udk[0]
+	v.udk = v.udk[1:]
+	v.dmu.Unlock()
+	return d
+}
+
+func (v *vqueue) flushAndLoadInsert() (uii []index) {
+	v.imu.Lock()
+	uii = make([]index, len(v.uii))
+	copy(uii, v.uii)
+	v.uii = v.uii[:0]
+	v.imu.Unlock()
+	sort.Slice(uii, func(i, j int) bool {
+		return uii[i].date > uii[j].date
+	})
+	dup := make(map[string]bool, len(uii)/2)
+	dl := make([]int, 0, len(uii)/2)
+	for i, idx := range uii {
+		v.imu.Lock()
+		delete(v.uiil, idx.uuid)
+		v.imu.Unlock()
+		if dup[idx.uuid] {
+			dl = append(dl, i)
+		} else {
+			dup[idx.uuid] = true
+		}
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(dl)))
+	for _, i := range dl {
+		uii = append(uii[:i], uii[i+1:]...)
+	}
+	sort.Slice(uii, func(i, j int) bool {
+		return uii[i].date < uii[j].date
+	})
+	return uii
+}
+
+func (v *vqueue) flushAndLoadDelete() (udk []key) {
+	v.dmu.Lock()
+	udk = make([]key, len(v.udk))
+	copy(udk, v.udk)
+	v.udk = v.udk[:0]
+	v.dmu.Unlock()
+	sort.Slice(udk, func(i, j int) bool {
+		return udk[i].date > udk[j].date
+	})
+	dup := make(map[string]bool, len(udk)/2)
+	dl := make([]int, 0, len(udk)/2)
+	for i, idx := range udk {
+		if dup[idx.uuid] {
+			dl = append(dl, i)
+		} else {
+			dup[idx.uuid] = true
+		}
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(dl)))
+	for _, i := range dl {
+		udk = append(udk[:i], udk[i+1:]...)
+	}
+	sort.Slice(udk, func(i, j int) bool {
+		return udk[i].date < udk[j].date
+	})
+
+	udm := make(map[string]int64, len(udk))
+	for _, d := range udk {
+		udm[d.uuid] = d.date
+	}
+	dl = dl[:0]
+	// we should check insert vqueue if insert vqueue exists and delete operation date is newer than insert operation date then we should remove insert vqueue's data.
+	v.imu.Lock()
+	for i, idx := range v.uii {
+		// check same uuid & operation date
+		// if date is equal, it may update operation we shouldn't remove at that time
+		date, exists := udm[idx.uuid]
+		if exists && date > idx.date {
```

Comment:
I'm not sure but is it needed to remove `idx.uuid` entry from `udk` here?

Reply:
Good question.
In the CreateIndex operation of the NGT Service, the Delete Queue is processed first, and then the Insert Queue is processed, so the Insert Queue still contains the old Insert Operation older than the Delete Queue, and it is possible that data that was intended to be deleted is registered again.
For this reason, the data is deleted from the Insert Queue only when retrieving data from the Delete Queue. 
In the case of old implementation called VCache, the opposite process (delete Delete queue when retrieving Insert Queue)  was also performed, but this time it is not performed because there were many problems in update operations such as Update if Delete is deleted.

I'll add some comment about it.

Reply:
I know about that, so I just wondered that `udk` will have an unexisting key after removing insert vqueue entry by `dl` index. I asked is it okay that `udk` is not updated here.

Reply:
sorry I don't get it, do you mean should I remove like `delete(udk, idx.uuid)` ????

Reply:
right.
Doesn't it make sense?

Reply:
No, it doesn't. If you delete the udk entry here, it will cause deletion omissions, and the delete instruction should pass through all of them, as it did in the vcahce bug mentioned above.
If there is a newer Insert operation than the Delete instruction, the Delete instruction should also be executed.

Reply:
I got. I've misunderstood because I forgot to think about L279-287 that removes duplicated entries in Delete Vqueue.

Reply:
I see, sorry for really complex code,,, I'd like to refactor in the far future... :sweat_smile: